### PR TITLE
feat: added dap ui and relative config

### LINKED ISSUE #3127: [hacktoberfest] Add DAP-UI

### PR DESCRIPTION
**Pull request by GabMus**
# Description

Added DAP UI, a default configuration adapted to lvim style and `<leader>dU` shortcut to toggle it.

fixes #3127

### CODE REVIEW

File: lua/lvim/core/dap.lua [hunk 1]
```diff
@@ -75,4 +76,62 @@ end
...
+      expand = { "<CR>", "<2-LeftMouse>" },
+      open = "o",
+      remove = "d",
+      edit = "e",
+      repl = "r",
+      toggle = "t",
+    },
+    -- Expand lines larger than the window
+    -- Requires >= 0.7
+    expand_lines = vim.fn.has "nvim-0.7",
```
Comment (lvimuser) [inline lua/lvim/core/dap.lua:93]:
```suggestion
    expand_lines = true,
```
0.7.2 is already a requirement, (and 0.8 for the next version )

**Comment by lvimuser** [comment]
Did you consider this [part](https://github.com/rcarriga/nvim-dap-ui#usage)?
> You can use nvim-dap events to open and close the windows automatically (:help dap-extensions)

So if you start/stop debugging the UI will open/close automatically. I think this would be the expected behavior for most users (and would aid discoverability), but maybe it's controversial? cc @ChristianChiarulli @kylo252 @abzcoding 

Would not hurt to make it optional I think, perhaps a `lvim.builtin.dap.ui.<flag>`.

**Review by lvimuser — APPROVED** [review]
LGTM

**Comment by ChristianChiarulli** [comment]
for now formatting is failing

**Review by lvimuser — CHANGES_REQUESTED** [review]
We should also update the highlight groups
```lua
breakpoint = {
  text = lvim.icons.ui.Bug,
  texthl = "DiagnosticSignError",
  linehl = "",
  numhl = "",
},
breakpoint_rejected = {
  text = lvim.icons.ui.Bug,
  texthl = "DiagnosticSignHint",
  linehl = "",
  numhl = "",
},
stopped = {
  text = lvim.icons.ui.BoldArrowRight,
  texthl = "DiagnosticSignInfo",
  linehl = "DiagnosticUnderlineInfo",
  numhl = "DiagnosticSignInfo",
},
```

**Comment by ChristianChiarulli** [comment]
@GabMus 

this is what we came up with after the stream using what you did as a base:

```local M = {}

M.config = function()
  lvim.builtin.dap = {
    active = false,
    on_config_done = nil,
    breakpoint = {
      text = lvim.icons.ui.Bug,
      texthl = "DiagnosticSignError",
      linehl = "",
      numhl = "",
    },
    breakpoint_rejected = {
      text = lvim.icons.ui.Bug,
      texthl = "LspDiagnosticsSignHint",
      linehl = "",
      numhl = "",
    },
    stopped = {
      text = lvim.icons.ui.BoldArrowRight,
      texthl = "DiagnosticSignWarn",
      linehl = "Visual",
      numhl = "DiagnosticSignWarn",
    },
    ui = {
      auto_open = true,
    },
  }
end

M.setup = function()
  local dap = require "dap"

  if lvim.use_icons then
    vim.fn.sign_define("DapBreakpoint", lvim.builtin.dap.breakpoint)
    vim.fn.sign_define("DapBreakpointRejected", lvim.builtin.dap.breakpoint_rejected)
    vim.fn.sign_define("DapStopped", lvim.builtin.dap.stopped)
  end

  lvim.builtin.which_key.mappings["d"] = {
    name = "Debug",
    t = { "<cmd>lua require'dap'.toggle_breakpoint()<cr>", "Toggle Breakpoint" },
    b = { "<cmd>lua require'dap'.step_back()<cr>", "Step Back" },
    c = { "<cmd>lua require'dap'.continue()<cr>", "Continue" },
    C = { "<cmd>lua require'dap'.run_to_cursor()<cr>", "Run To Cursor" },
    d = { "<cmd>lua require'dap'.disconnect()<cr>", "Disconnect" },
    g = { "<cmd>lua require'dap'.session()<cr>", "Get Session" },
    i = { "<cmd>lua require'dap'.step_into()<cr>", "Step Into" },
    o = { "<cmd>lua require'dap'.step_over()<cr>", "Step Over" },
    u = { "<cmd>lua require'dap'.step_out()<cr>", "Step Out" },
    p = { "<cmd>lua require'dap'.pause()<cr>", "Pause" },
    r = { "<cmd>lua require'dap'.repl.toggle()<cr>", "Toggle Repl" },
    s = { "<cmd>lua require'dap'.continue()<cr>", "Start" },
    q = { "<cmd>lua require'dap'.close()<cr>", "Quit" },
    U = { "<cmd>lua require'dapui'.toggle()<cr>", "Toggle UI" },
  }

  if lvim.builtin.dap.on_config_done then
    lvim.builtin.dap.on_config_done(dap)
  end
end

M.setup_ui = function()
  local dap = require "dap"
  local dapui = require "dapui"
  dapui.setup {

    expand_lines = true,
    icons = {
      expanded = "",
      collapsed = "",
      circular = "",
    },
    layouts = {
      {
        elements = {
          { id = "scopes", size = 0.33 },
          { id = "breakpoints", size = 0.17 },
          { id = "stacks", size = 0.25 },
          { id = "watches", size = 0.25 },
        },
        size = 0.33,
        position = "right",
      },
      {
        elements = { { id = "repl", size = 0.45 }, { id = "console", size = 0.55 } },
        size = 0.27,
        position = "bottom",
      },
    },
    floating = {
      max_width = 0.9,
      max_height = 0.5,
      border = vim.g.border_chars,
      mappings = {
        close = { "q", "<Esc>" },
      },
    },
  }

  if lvim.builtin.dap.ui.auto_open then
    dap.listeners.after.event_initialized["dapui_config"] = function()
      dapui.open()
    end
    -- dap.listeners.before.event_terminated["dapui_config"] = function()
    --   dapui.close()
    -- end
    -- dap.listeners.before.event_exited["dapui_config"] = function()
    --   dapui.close()
    -- end
  end
end

return M
```

**Review by ChristianChiarulli — APPROVED** [review]

**Comment by hawkinchina** [comment]
can we change the border style of dap-ui ?
image

**Pull request merged**